src/ripple/consensus/LedgerTrie.h

```diff
@@ -0,0 +1,712 @@
+        };
+
+        // Return a span of this over the half-open interval [from,to)
+        Span
+        sub(Seq from, Seq to)
+        {
+            return Span(clamp(from), clamp(to), ledger_);
+        }
+
+        friend std::ostream&
+        operator<<(std::ostream& o, Span const& s)
+        {
+            return o << s.tip().second << "(" << s.start_ << "," << s.end_
+                     << "]";
+        }
+
+        friend Span
+        merge(Span const& a, Span const& b)
+        {
+            // Return combined span, using ledger_ from higher sequence span
+            if (a.end_ < b.end_)
+                return Span(std::min(a.start_, b.start_), b.end_, b.ledger_);
+
+            return Span(std::min(a.start_, b.start_), a.end_, a.ledger_);
+        }
+    };
+
+    // A node in the trie
+    struct Node
+    {
+        Node() : span{}, tipSupport{0}, branchSupport{0}
+        {
+        }
+
+        Node(Ledger const& l) : span{l}, tipSupport{1}, branchSupport{1}
+        {
+        }
+
+        Node(Span s) : span{std::move(s)}
+        {
+        }
+
+        Span span;
+        std::uint32_t tipSupport = 0;
+        std::uint32_t branchSupport = 0;
+
+        std::vector<std::unique_ptr<Node>> children;
+        Node* parent = nullptr;
+
+        /** Remove the given node from this Node's children
+
+            @param child The address of the child node to remove
+            @note The child must be a member of the vector. The passed pointer
+                  will be dangling as a result of this call
+        */
+        void
+        erase(Node const* child)
+        {
+            auto it = std::remove_if(
+                    children.begin(),
+                    children.end(),
+                    [child](std::unique_ptr<Node> const& curr) {
+                        return curr.get() == child;
+                    });
+            assert(it != children.end());
+            children.erase(it, children.end());
+        }
+
+        friend std::ostream&
+        operator<<(std::ostream& o, Node const& s)
+        {
+            return o << s.span << "(T:" << s.tipSupport
+                     << ",B:" << s.branchSupport << ")";
+        }
+
+        Json::Value
+        getJson() const
+        {
+            Json::Value res;
+            res["id"] = to_string(span.tip().second);
+            res["seq"] = static_cast<std::uint32_t>(span.tip().first);
+            res["tipSupport"] = tipSupport;
+            res["branchSupport"] = branchSupport;
+            if(!children.empty())
+            {
+                Json::Value &cs = (res["children"] = Json::arrayValue);
+                for(auto const & child : children)
+                {
+                    cs.append(child->getJson());
+                }
+            }
+            return res;
+        }
+    };
+
+    // The root of the trie. The root is allowed to break the no-single child
+    // invariant.
+    std::unique_ptr<Node> root;
```

Comment:
Instead of `unique_ptr` can this be a value type instead (i.e. `Node`)

Reply:
I considered that, but I felt the uniformity of all `Node`s accessed via a pointer like interface trumped the benefit of a value type.